File: src/main/java/net/jgp/books/spark/ch14/lab920_column_as_parameter/ColumnAdditionApp.java
package net.jgp.books.spark.ch14.lab920_column_as_parameter;

import static org.apache.spark.sql.functions.callUDF;
import static org.apache.spark.sql.functions.col;
import static org.apache.spark.sql.functions.array;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

/**
 * Additions via UDF.
 * 
 * @author jgp
 */
public class ColumnAdditionApp {

  private static final int COL_COUNT = 8;

  /**
   * main() is your entry point to the application.
   * 
   * @param args
   */
  public static void main(String[] args) {
    ColumnAdditionApp app = new ColumnAdditionApp();
    app.start();
  }

  /**
   * The processing code.
   */
  private void start() {
    // Creates a session on a local master
    SparkSession spark = SparkSession.builder()
        .appName("Column addition")
        .master("local[*]")
        .getOrCreate();
    spark.udf().register(
        "add",
        new ColumnAdditionUdf(),
        DataTypes.IntegerType);

    Dataset<Row> df = createDataframe(spark);

    Column[] cols = new Column[COL_COUNT];
    for (int i = 0; i < COL_COUNT; i++) {
      cols[i] = df.col("c" + i);
    }
    Column col = array(cols);

    df = df
        .withColumn(
            "sum",
            callUDF("add", col));
    df.show(false);

  }

  private static Dataset<Row> createDataframe(SparkSession spark) {
    StructType schema = DataTypes.createStructType(new StructField[] {
        DataTypes.createStructField(
            "c0",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c1",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c2",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c3",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c4",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c5",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c6",
            DataTypes.IntegerType,
            false),
        DataTypes.createStructField(
            "c7",
            DataTypes.IntegerType,
            false) });

    List<Row> rows = new ArrayList<>();
    rows.add(RowFactory.create(1, 2, 4, 8, 16, 32, 64, 128));
    rows.add(RowFactory.create(0, 0, 0, 0, 0, 0, 0, 0));
    rows.add(RowFactory.create(1, 1, 1, 1, 1, 1, 1, 1));

    return spark.createDataFrame(rows, schema);
  }
}

File: src/main/python/lab920_column_as_parameter/columnAdditionApp.py
"""
 Column Additions via UDF.

 @author rambabu.posa
"""
from pyspark.sql import (SparkSession, functions as F)
from pyspark.sql.types import (StructType,StructField, IntegerType)

def createDataframe(spark: SparkSession):
    schema = StructType([
        StructField('c0', IntegerType(), False),
        StructField('c1', IntegerType(), False),
        StructField('c2', IntegerType(), False),
        StructField('c3', IntegerType(), False),
        StructField('c4', IntegerType(), False),
        StructField('c5', IntegerType(), False),
        StructField('c6', IntegerType(), False),
        StructField('c7', IntegerType(), False)
    ])

    rows = [
        (1, 2, 4, 8,16, 32, 64, 128),
        (0, 0, 0, 0,0, 0, 0, 0),
        (1, 1, 1, 1,1, 1, 1, 1)
    ]
    return spark.createDataFrame(rows, schema)

def column_addition(col):
    res = 0
    for value in col:
        res = res + value

    return res

def main(spark):
    df = createDataframe(spark)
    df.show(truncate=False)

    add_col = F.udf(column_addition, IntegerType())

    cols =[df.c0, df.c1, df.c2, df.c3,
           df.c4, df.c5, df.c6, df.c7]

    col = F.array(*cols)

    df = df.withColumn("sum", add_col(col))

    df.show(truncate=False)


if __name__ == '__main__':
    # Creates a session on a local master
    spark = SparkSession.builder.appName("Column addition") \
        .master("local[*]").getOrCreate()
    # Comment this line to see full log
    spark.sparkContext.setLogLevel('error')
    main(spark)
    spark.stop()